### PR TITLE
fix: make URL domain detection linear

### PR DESCRIPTION
**Pull request by seratch**
## Problem

The URL Filter used a backtracking regular expression to detect scheme-less domains. Inputs with long dotted runs that never ended in a valid top-level domain caused quadratic matching time, allowing untrusted text to block the synchronous guardrail path for a disproportionate amount of time.

## Solution

Replace the backtracking domain expression with a linear-time scanner that first identifies maximal hostname-shaped candidates and then locates the last valid ASCII domain suffix in a single bounded pass. The scanner preserves the previous matching behavior, including partial matches and path handling, while avoiding repeated suffix backtracking.

Add regression coverage for existing edge cases, property-based equivalence against the previous expression on bounded inputs, and a large invalid dotted-input performance case.

### CODE REVIEW

File: tests/unit/checks/test_urls.py [hunk 1]
```diff
@@ -2,16 +2,84 @@
...
+def test_detect_urls_handles_large_invalid_domain_input_quickly() -> None:
+    """Invalid dotted input should not cause quadratic regex backtracking."""
+    text = "a." * 16_000 + "-"
+
+    started_at = perf_counter()
+    detected = _detect_urls(text)
+    duration_seconds = perf_counter() - started_at
+
+    assert detected == []  # noqa: S101
+    assert duration_seconds < 1.0  # noqa: S101
```
[the remark this answers is not in the document]
Comment (seratch) [reply tests/unit/checks/test_urls.py]: @codex review
